src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs

```diff
@@ -236,27 +230,40 @@ public void BeginListen()
                             });
                         }
                     }
-                    catch (OperationCanceledException)
-                    {
-                    }
-                    catch (Exception ex)
-                    {
-                        unhandledException = ex;
-                    }
-                    NodeServerTrace.Information("Stop listening");
-                    this.EndListen(unhandledException);
                 }
+                catch (OperationCanceledException)
+                {
+                    this.logger.LogTrace("Listening cancelled.");
+                }
+                catch (Exception ex)
+                {
+                    this.logger.LogTrace("Exception occurred: {0}", ex);
+                    unhandledException = ex;
+                }
+
+                this.logger.LogDebug("Terminating listening thread.");
+                this.EndListen(unhandledException);
+
+                this.logger.LogTrace("(-)");
             }).Start();
+
+            this.logger.LogTrace("(-)");
         }
 
         private void EndListen(Exception unhandledException)
         {
+            this.logger.LogTrace("()");
+
             if (Interlocked.CompareExchange(ref this.cleaningUp, 1, 0) == 1)
+            {
+                this.logger.LogTrace("(-)[CLEANING_UP]");
```

Comment:
Should this be `NOT_CLEANING_UP`

Reply:
consider two such guards
```
if (x)
{
   exit
}

if (y)
{
  exit
}

actually do cleanup
```

with the proposed logic, you would put in both X and Y branches `trace(not cleaning up)`, but the purpose of the context is to distinguish which one actually executed

so rather than using `trace(not cleaning up)` we want something like `trace(exit because of x)` and  `trace(exit because of y)`

so we are existing here because cleanup is already in progress, thus CLEANING_UP

could be ALREADY_IN_PROGRESS or ALREADY_CLEANING or anything else, but NOT_CLEANING_UP would make less sense to me